File: Python/Graphs/GraphTraversal.py
class Graph:
    def __init__(self):
        self.adjacencyList = {}

    def addVertex(self, v):
        if v not in self.adjacencyList:
            self.adjacencyList[v] = []

    def addEdge(self, v1, v2):
        if v1 not in self.adjacencyList:
            self.addVertex(v1)
        if v2 not in self.adjacencyList:
            self.addVertex(v2)
        self.adjacencyList[v1].append(v2)
        self.adjacencyList[v2].append(v1)

    def depthFirstIterative(self, vertex):
        output = []
        stack = []
        visited = {}
        if vertex not in self.adjacencyList or len(self.adjacencyList) == 0:
            return
        stack.append(vertex)
        visited[vertex] = True
        while len(stack) != 0:
            vertex = stack.pop()
            output.append(vertex)
            for v in self.adjacencyList[vertex]:
                if v not in visited:
                    stack.append(v)
                    visited[v] = True
        return output

    def depthFirstRecursive(self, vertex):
        output = []
        visited = {}
        def helper(vertex):
            if vertex not in self.adjacencyList or len(self.adjacencyList) == 0:
                return
            visited[vertex] = True
            output.append(vertex)
            for v in self.adjacencyList[vertex]:
                if v not in visited:
                    helper(v)
        helper(vertex)
        return output

    def breadthFristSearch(self, vertex):
        output = []
        visited = {}
        queue = []
        queue.append(vertex)
        visited[vertex] = True
        while len(queue) != 0:
            vertex = queue.pop(0)
            output.append(vertex)
            for v in self.adjacencyList[vertex]:
                if v not in visited:
                    visited[v] = True
                    queue.append(v)
        return output

g = Graph()
g.addVertex('Tokyo')
g.addVertex('Cairo')
g.addVertex('Dallas')
g.addEdge('Tokyo', 'Cairo')
g.addEdge('Dallas', 'Tokyo')
g.addVertex('Hyderabad')
g.addEdge('Hyderabad', 'Cairo')
g.addEdge('Dallas', 'Hyderabad')
g.addEdge('Chicago', 'Mason')
g.addEdge('Dallas', 'Mason')
print(g.depthFirstIterative('Tokyo'))
print(g.depthFirstRecursive('Tokyo'))
print(g.breadthFristSearch('Tokyo'))

File: JavaScript/Graphs/GraphTraversal.js
class Graph {
    constructor() {
        this.adjacencyList = {};
    }

    addVertex(v1) {
        if (!this.adjacencyList[v1]) {
            this.adjacencyList[v1] = [];
        }
    }

    addEdge(v1, v2) {
        if (this.adjacencyList[v1]) {
            this.adjacencyList[v1].push(v2);
        }
        if (this.adjacencyList[v2]) {
            this.adjacencyList[v2].push(v1);
        }
    }

    depthFirstRecursive(vertex) {
        let output = [];
        let visited = {};
        let list = this.adjacencyList;
        function helper(vertex) {
            if (!list[vertex] || list[vertex].length === 0) {
                return;
            } else {
                visited[vertex] = true;
                output.push(vertex);
                for (let i = 0; i < list[vertex].length; i++) {
                    if (!visited[list[vertex][i]]) {
                        helper(list[vertex][i]);
                    }
                }
            }
        }
        helper(vertex);
        return output;
    }

    depthFirstIterative(vertex) {
        let output = [];
        let stack = [];
        let visited = {};
        let currentVertex;

        stack.push(vertex);
        visited[vertex] = true;
        while (stack.length !== 0) {
            currentVertex = stack.pop();
            output.push(currentVertex);
            this.adjacencyList[currentVertex].forEach(element => {
                if (!visited[element]) {
                    stack.push(element);
                    visited[element] = true;
                }
            });
        }
        return output;
    }

    breadthFirstTraversal(vertex) {
        let output = [];
        let visited = {};
        let queue = [];
        let currentVertex;

        queue.push(vertex);
        while (queue.length !== 0) {
            currentVertex = queue.shift();
            if (!visited[currentVertex]) {
                visited[currentVertex] = true;
                output.push(currentVertex);
                queue.push(...this.adjacencyList[currentVertex]);
            }
        }
        return output;
    }
}

let g = new Graph();
g.addVertex("A");
g.addVertex("B");
g.addVertex("C");
g.addVertex("D");
g.addVertex("E");
g.addVertex("F");
g.addVertex("G");
g.addEdge("A", "B");
g.addEdge("A", "C");
g.addEdge("B", "D");
g.addEdge("C", "E");
g.addEdge("D", "E");
g.addEdge("D", "F");
g.addEdge("E", "F");
console.log(g);
console.log(g.depthFirstRecursive("A"));
console.log(g.depthFirstIterative("A"));
console.log(g.breadthFirstTraversal("A"));